ZCS-15087 : Forceful usage counter : Modify Cos Request

swathy-shaji opened the pull request:
Ticket - [ZCS-15087](https://synacor.atlassian.net/browse/ZCS-15087)

[ZCS-15087]: https://synacor.atlassian.net/browse/ZCS-15087?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

swathy-shaji commented:
@RakeshAMore , 

Need to add new attributes in the ModifyCos response to send the statuscode and status as we discussed earlier
Please recheck the countObjects method written not sure the approach is good. It is written in such a way as to make it reusable for further conditions as Chitransu also needs to write a new filter for his ticket and can be handled under the same method. Please refactor it if it does not good

RakeshAMore commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -10538,4 +10538,8 @@ TODO: delete them permanently from here
   <desc>Zimbra feature actual usage count</desc>
 </attr>
 
+<attr id="4134" name="zimbraLicenseModifyCosFailedAttributes" type="string" cardinality="single" optionalIn="cos" since="10.1.0">
```
I gues we dont need thsi attribute now

swathy-shaji replied:
Its already removed

RakeshAMore commented on soap/src/java/com/zimbra/soap/admin/message/ModifyCosResponse.java:
```diff
@@ -38,6 +38,12 @@ public class ModifyCosResponse {
     @XmlElement(name=AdminConstants.E_COS)
     private CosInfo cos;
 
+    @XmlAttribute(name=AdminConstants.E_CODE)
```
As per the current flow we dont require this now.

swathy-shaji replied:
Its already removed

log2akshat commented on store/src/java/com/zimbra/cs/account/ldap/LdapProvisioning.java:
```diff
@@ -492,6 +491,38 @@ public void modifyAttrs(Entry e, Map<String, ? extends Object> attrs,
+     */
+    private void postModificationValidation(Entry e, Map<String, ?> attrs) throws ServiceException {
+        if (e instanceof Account && !(e instanceof CalendarResource)) {
+            Account acct = (Account) e;
```
Please reduce the redundant checks, type checking for `Account` and `CalendarResource` can be combined into one if statement which will improve readability, provide clearer separation of conditions, and offer more flexibility for future changes.

swathy-shaji replied:
Done

log2akshat commented on store/src/java/com/zimbra/cs/account/ldap/LdapProvisioning.java:
```diff
@@ -9451,6 +9483,45 @@ public long countObjects(CountObjectsType type, Domain domain, UCService ucServi
 
+    @Override
+    public long countObjects(CountObjectsType type, Object... args) throws ServiceException {
+        ZLdapFilter filter = null;
```
Please add null checks to ensure that the arguments passed are not null and are of the expected types and quantities.

swathy-shaji replied:
Done

log2akshat commented on store/src/java/com/zimbra/cs/account/ldap/LdapProvisioning.java:
```diff
@@ -9451,6 +9483,45 @@ public long countObjects(CountObjectsType type, Domain domain, UCService ucServi
+            case internalUserAccountsWithLdapFeatureCheck: // accounts with a feature enabled or disabled
+                ldapAttribute = (String) args[0];
+                String ldapValue = (String) args[1];
+                types.add(ObjectType.accounts);
```
Avoid redundancy in the code by moving out `ObjectType.accounts` to `types` set, out of the `switch` block.

swathy-shaji replied:
here we cannot move this out of the switch block, as it can vary on new upcoming cases. for now whatever the cases we have all are using same type. but there is a possibility of changing it on new case.

log2akshat commented on store/src/java/com/zimbra/cs/ldap/unboundid/UBIDLdapFilterFactory.java:
```diff
@@ -860,6 +863,60 @@ public ZLdapFilter accountsWithDocumentEditing() {
+        try {
+            cos = Provisioning.getInstance().get(Key.CosBy.id, cosId);
+        } catch (Exception e) {
+            ZimbraLog.system.warn("unable to fetch cos " + cosId);
```
Do not use String concatenation while displaying message as it will create a new String in the heap memory.

swathy-shaji replied:
Done

log2akshat commented on store/src/java/com/zimbra/cs/ldap/unboundid/UBIDLdapFilterFactory.java:
```diff
@@ -860,6 +863,60 @@ public ZLdapFilter accountsWithDocumentEditing() {
+                                        Filter.createEqualityFilter(ldapAttribute, LdapConstants.LDAP_TRUE),
+                                        Filter.createEqualityFilter(ldapAttribute, LdapConstants.LDAP_FALSE))),
+                        Filter.createORFilter(filters)));
+    }
```
Please create a  separate centralized `Filter` method for the common logic for creating the main LDAP filter which will centralize the filter creation, making the code `DRY (Don't Repeat Yourself)` and easier to maintain.

swathy-shaji replied:
Here we are already using combined filter wherever it is possible. For the Ldap check and cos check. the filters are dependent on the dynamic values for which we can't create centralized filters

log2akshat commented on store/src/java/com/zimbra/cs/ldap/unboundid/UBIDLdapFilterFactory.java:
```diff
@@ -860,6 +863,60 @@ public ZLdapFilter accountsWithDocumentEditing() {
+                                Filter.createEqualityFilter(ldapAttribute, LdapConstants.LDAP_TRUE),
+                                Filter.createEqualityFilter(ldapAttribute, LdapConstants.LDAP_FALSE))),
+                        Filter.createORFilter(filters)));
+    }
```
Use  `stream` for creating filters in `accountsByCosesAndFeatureCheck` to reduces boilerplate code and improved readability.

swathy-shaji replied:
Done

log2akshat reviewed (changes requested):
Also, while raising the Pull Request please follow the proper Zimbra standards in the following format:

Problem Statement
Approach and Fix
Testing Done
Links to any other related pull requests or resources.

swathy-shaji committed:
ZCS-15087: Added new LDAP filters for licensing

swathy-shaji commented:
Closing this PR since we opened a new PR with the same changes as we created a new branch for the same due to which we are releasing these changes as a part of different ticket 

We addressed all the comments and considered the same in to the other PR as well. Please check

https://github.com/Zimbra/zm-mailbox/pull/1618
